data/src/frame.rs

```diff
@@ -423,106 +419,90 @@ impl DefaultFrameBuffer {
     }
 }
 
-pub type ArcFrame = Arc<Frame>;
+/// Decoded frame information.
+#[derive(Debug)]
+pub struct Frame {
+    /// The kind of frame (audio or video).
+    pub kind: MediaKind,
+    /// Frame buffer containing the data associated to each plane.
+    pub buf: Box<dyn FrameBuffer>,
+    /// Timestamp information associated to a frame.
+    pub t: TimeInfo,
+}
 
-pub fn new_default_frame<T>(kind: T, t: Option<TimeInfo>) -> Frame
-where
-    T: Into<MediaKind> + Clone,
-{
-    let k = kind.into();
-    let buf = DefaultFrameBuffer::new(&k);
+impl Frame {
+    /// Creates a new frame.
+    pub fn new_default_frame<T>(kind: T, t: Option<TimeInfo>) -> Self
+    where
+        T: Into<MediaKind> + Clone,
+    {
+        let k = kind.into();
+        let buf = DefaultFrameBuffer::new(&k);
 
-    Frame {
-        kind: k,
-        buf: Box::new(buf),
-        t: t.unwrap_or_default(),
+        Self {
+            kind: k,
+            buf: Box::new(buf),
+            t: t.unwrap_or_default(),
+        }
     }
-}
 
-fn copy_plane(
-    dst: &mut [u8],
-    dst_linesize: usize,
-    src: &[u8],
-    src_linesize: usize,
-    w: usize,
-    h: usize,
-) {
-    let dst_chunks = dst.chunks_mut(dst_linesize);
-    let src_chunks = src.chunks(src_linesize);
+    /// Copies a determined video plane in an output buffer.
+    pub fn copy_video_plane(
+        &self,
+        index: usize,
+        dst: &mut [u8],
+        dst_linesize: usize,
+        width: usize,
+        height: usize,
+    ) -> Result<(), FrameError> {
+        if matches!(self.kind, MediaKind::Video(_)) {
+            let src = self.buf.as_slice_inner(index).unwrap();
+            let src_linesize = self.buf.linesize(index).unwrap();
 
-    for (d, s) in dst_chunks.zip(src_chunks).take(h) {
-        unsafe {
-            copy_nonoverlapping(s.as_ptr(), d.as_mut_ptr(), w);
+            copy_plane(dst, dst_linesize, src, src_linesize, width, height);
+
+            Ok(())
+        } else {
+            Err(FrameError::InvalidKind)
         }
     }
-}
 
-fn copy_image<'a, IM, IU, I>(
-    dst: IM,
-    dst_linesizes: IU,
-    src: I,
-    src_linesizes: IU,
-    w: usize,
-    h: usize,
-    fmt: &Formaton,
-) where
-    IM: Iterator<Item = &'a mut [u8]>,
-    I: Iterator<Item = &'a [u8]>,
-    IU: Iterator<Item = usize>,
-{
-    let dst_iter = dst.zip(dst_linesizes);
-    let src_iter = src.zip(src_linesizes);
-    let iter = dst_iter.zip(src_iter).zip(fmt.iter());
-
-    for (((d, d_linesize), (s, s_linesize)), c) in iter {
-        copy_plane(
-            d,
-            d_linesize,
-            s,
-            s_linesize,
-            c.unwrap().get_width(w),
-            c.unwrap().get_height(h),
-        );
-    }
-}
+    /// Copies a single video frame in an output buffer.
+    pub fn copy_video_frame<'a, IM, IU>(
```

Comment:
I'm not sure we have to make it pub.

Reply:
Is there a particular reason?

Reply:
A possible solution would be run `copy_video_plane` per the number of planes and save them into n_planes different buffers, then merge all of them into a single one. A bit more cumbersome for the user perhaps, but at least we remove an API :)

Reply:
mediakind-specific API look quite bad on the Frame.